File: python/life/param.py
class Param:
    
    def __init__(self):
        self.width = 128
        self.height = 64
        self.birthValues = {
            "b3": 1,
        }
        self.surviveValues = {
            "s2": 1,
            "s3": 1,
        }
        self.spawnPercent = 20
        self.wrapX = True
        self.wrapY = True


"""
REGEXP patterns edited to prevent python warning!!!


    {
        try {
            $this->parse_ini($section, $rules);
        } catch (ErrorProcessingINI $e) {
            echo "Unable to read ini file: ".$e->getMessage()."\n";
        }
    }

    private function parse_ini($section, $rules) {
        $sim_ini = "../game-of-life.ini";
        if (!file_exists($sim_ini)) {
            throw new ErrorProcessingINI("File not found: '$sim_ini'");
        }
        $ini = parse_ini_file($sim_ini, true);

        $section = "world-$section";
        if (!array_key_exists($section, $ini)) {
            throw new ErrorProcessingINI("Section not found: '$section'");
        }
        if (!array_key_exists("rules", $ini)) {
            throw new ErrorProcessingINI("Section not found: 'rules'");
        }
        $setup = $ini[$section];
        $ruleset = $ini["rules"];

        $keys = array("resolution", "wrapx", "wrapy", "spawn_percent");    // "grid" not used yet
        foreach ($keys as $key) {
            if (!array_key_exists($key, $setup)) {
                throw new ErrorProcessingINI("Value '$key' not found in $section section");
            }
        }

        if (preg_match("/^(\\d+)x(\\d+)$/",$setup["resolution"],$res)) {
            $this->width = $res[1];
            $this->height = $res[2];
        } else {
            throw new ErrorProcessingINI("Expected ##x## in 'resolution'");
        }

        $this->wrapx = ($setup["wrapx"] === "yes");
        $this->wrapy = ($setup["wrapy"] === "yes");
        $this->spawn_percent = intval($setup["spawn_percent"]);
        if ($this->spawn_percent < 5 || $this->spawn_percent > 95) {
            throw new ErrorProcessingINI("Expected positive numeric spawn_percent (between 5 and 95)");
        }

        if (preg_match("/^b(\\d+)\\//",$ruleset[$rules],$birth)) {
            foreach (preg_split("//",$birth[1]) as $val) {
                IF ($val <> "") {
                    $this->birth_values["b$val"] = 1;
                }
            }
        } else {
            throw new ErrorProcessingINI("Birth rule not specified; expecting 'b#/...'");
        }
        if (preg_match("/\\/s(\\d+)$/",$ruleset[$rules],$survive)) {
            foreach (preg_split("//",$survive[1]) as $val) {
                if ($val <> "") {
                    $this->survive_values["s$val"] = 1;
                }
            }
        } else {
            throw new ErrorProcessingINI("Survival rule not specified; expecting '.../s#'");
        }

    }
}
"""

File: go/life/param.go
package life

import (
	"log"
	"regexp"
	"strconv"
	"strings"
)

// Param provides interface to values from ini file
type Param struct {
	Width, Height int
	RuleValues    map[string]map[string]bool
	SpawnPercent  int
	WrapX, WrapY  bool
	ini           INI
}

// Init sets values to those from ini file, or to defaults if an error occurs
func (s *Param) Init(board string, rules string) {
	s.RuleValues = make(map[string]map[string]bool)
	s.RuleValues["b"] = make(map[string]bool)
	s.RuleValues["s"] = make(map[string]bool)

	s.setDefaults()
	err := s.ini.Parse("../game-of-life.ini")
	if err != nil {
		log.Print("world.Init: Error reading INI file: " + err.Error())
		log.Print("world.Init: using Default parameters instead")
		return
	}

	section := "world-" + board
	_, err = s.ini.Section(section)
	if err != nil {
		log.Printf(err.Error())
		return
	}

	s.readResolution(section)
	s.readWrap(section)
	s.readSpawn(section)
	s.readRules(rules)
}

func (s *Param) readResolution(sectName string) {
	val, err := s.ini.Value(sectName, "resolution")
	if err == nil {
		s.setResolution(val)
	} else {
		log.Print(err.Error())
		log.Print("Param: resolution not found, using default")
	}
}

func (s *Param) readWrap(sectName string) {
	val1, err1 := s.ini.Value(sectName, "wrapx")
	val2, err2 := s.ini.Value(sectName, "wrapy")
	if err1 == nil && err2 == nil {
		s.setWrap(val1 == "yes", val2 == "yes")
	} else {
		log.Print("Param: unable to read wrap values, using defaults")
	}
}

func (s *Param) readSpawn(sectName string) {
	val, err1 := s.ini.Value(sectName, "spawn_percent")
	if err1 == nil {
		percent, err2 := strconv.Atoi(val)
		if err2 == nil {
			s.setSpawn(percent)
		} else {
			log.Printf("Param: spawn percentage '%s' not numeric, using default", val)
		}
	} else {
		log.Print("Param: unable to read spawn percentage, using default")
	}
}

func (s *Param) readRules(valName string) {
	val, err := s.ini.Value("rules", valName)
	if err == nil {
		s.setRules(val)
	} else {
		log.Printf("Param: could not load '%s' rules, using default (classic Conway)", valName)
	}
}

func (s *Param) setDefaults() {
	// world configuration "default", AKA my VZ-200 resolution
	s.setResolution("128x64")
	s.setWrap(true, true)
	s.setSpawn(20)

	// default rules, the "Conway" classics
	s.setRules("b3/s23")
}

func (s *Param) setResolution(resolution string) {
	reRes, err := regexp.Compile(`^(\d+)x(\d+)$`)
	if err != nil {
		log.Fatal(err)
	}
	s.Width, err = strconv.Atoi(reRes.FindStringSubmatch(resolution)[1])
	if err != nil {
		log.Fatal(err)
	}
	s.Height, err = strconv.Atoi(reRes.FindStringSubmatch(resolution)[2])
	if err != nil {
		log.Fatal(err)
	}
}

func (s *Param) setWrap(wrapX, wrapY bool) {
	s.WrapX = wrapX
	s.WrapY = wrapY
}

func (s *Param) setSpawn(percentage int) {
	s.SpawnPercent = percentage
}

func (s *Param) setRules(simRules string) {
	reRules, err := regexp.Compile(`b(\d*)/s(\d*)`)
	if err != nil {
		log.Fatal(err)
	}
	bDigits := reRules.FindStringSubmatch(simRules)[1]
	sDigits := reRules.FindStringSubmatch(simRules)[2]
	s.parseRuleDigits("b", bDigits)
	s.parseRuleDigits("s", sDigits)
}

func (s *Param) parseRuleDigits(rule string, digits string) {
	digitSlice := strings.Split(digits, "")
	for _, digit := range digitSlice {
		s.RuleValues[rule][digit] = true
	}
}
